src/main/scala/org/broadinstitute/dsde/rawls/model/JsonSupport.scala

```diff
@@ -1,57 +1,145 @@
+  //Magic strings we use in JSON serialization
 
-  implicit object AttributeFormat extends RootJsonFormat[Attribute] {
+  //Entity refs get serialized to e.g. { "entityType" : "sample", "entityName" : "theBestSample" }
+  val ENTITY_TYPE_KEY = "entityType"
+  val ENTITY_NAME_KEY = "entityName"
+  val ENTITY_OBJECT_KEYS = Set(ENTITY_TYPE_KEY, ENTITY_NAME_KEY)
 
-    override def write(obj: Attribute): JsValue = obj match {
+  trait AttributeFormat extends RootJsonFormat[Attribute] with AttributeListSerializer {
+
+    override def write(obj: Attribute): JsValue = writeAttribute(obj)
+    def writeAttribute(obj: Attribute): JsValue = obj match {
+      //vals
       case AttributeNull => JsNull
       case AttributeBoolean(b) => JsBoolean(b)
       case AttributeNumber(n) => JsNumber(n)
       case AttributeString(s) => JsString(s)
-      case AttributeValueList(l) => JsArray(l.map(write(_)):_*)
-      case AttributeEntityReferenceList(l) => JsArray(l.map(write(_)).toSeq:_*)
-      case AttributeEntityReference(entityType, entityName) => JsObject(Map("entityType" -> JsString(entityType), "entityName" -> JsString(entityName)))
-      case AttributeEmptyList => JsArray()
+      //ref
+      case AttributeEntityReference(entityType, entityName) => JsObject(Map(ENTITY_TYPE_KEY -> JsString(entityType), ENTITY_NAME_KEY -> JsString(entityName)))
+      //list types
+      case x: AttributeList[_] => writeListType(x)
+
+      case _ => throw new SerializationException("AttributeFormat doesn't know how to write JSON for type " + obj.getClass.getSimpleName)
     }
 
-    override def read(json: JsValue): Attribute = json match {
+    override def read(json: JsValue): Attribute = readAttribute(json)
+    def readAttribute(json: JsValue): Attribute = json match {
       case JsNull => AttributeNull
       case JsString(s) => AttributeString(s)
       case JsBoolean(b) => AttributeBoolean(b)
       case JsNumber(n) => AttributeNumber(n)
-      case JsArray(a) => getAttributeList(a.map(read(_)))
-      case JsObject(members) => AttributeEntityReference(members("entityType").asInstanceOf[JsString].value, members("entityName").asInstanceOf[JsString].value)
-      case _ => throw new DeserializationException("unexpected json type")
-    }
 
-    def getAttributeList(s: Seq[Attribute]) = s match {
-      case e: Seq[_] if e.isEmpty => AttributeEmptyList
-      case v: Seq[AttributeValue @unchecked] if (s.map(_.isInstanceOf[AttributeValue]).reduce(_&&_)) => AttributeValueList(v)
-      case r: Seq[AttributeEntityReference @unchecked] if (s.map(_.isInstanceOf[AttributeEntityReference]).reduce(_&&_)) => AttributeEntityReferenceList(r)
-      case _ => throw new DeserializationException("illegal array type")
+      case JsObject(members) if ENTITY_OBJECT_KEYS subsetOf members.keySet =>
+        AttributeEntityReference(members(ENTITY_TYPE_KEY).asInstanceOf[JsString].value, members(ENTITY_NAME_KEY).asInstanceOf[JsString].value)
```

Comment:
do a match instead of type casting:
(members(ENTITY_TYPE_KEY), members(ENTITY_NAME_KEY)) match {
  case (JsString(x), JsString(y)) =>
  case _ => throw exception
}